### PR TITLE
Yashwanth individual permission change show

### PR DESCRIPTION
**Pull request by yashwanth170**
# Description
The permission change log table only logs changes made to role, so I have added additional features to log permission changes made to individual account. Now the table shows latest logs of changes made to both individual and role

Bug_description

## Related PRS (if any):
This frontend PR is related to the [#1149](https://github.com/OneCommunityGlobal/HGNRest/pull/1149) backend PR.
To test this frontend PR you need to checkout the 'Yashwanth_individual_permission_change_show_backend' Backend PR.
…

## Main changes explained:
- When changes are made to indvidual or user, the function calls endpoint POST_USER_PERMISSION_CHANGE_LOGS to post changes made to permissions
- Created new endpoint GET_USER_PERMISSION_CHANGE_LOGS to fetch changes made to both individual and role

## How to test:
1. check into current branch, and check into backend branch Yashwanth_individual_permission_change_show_backend
2. do `npm install` and `...` to run this PR locally
3. Clear site data/cache
4. log as admin user
5. go to dashboard→ Other Links → Permission Management → Make changes to a role or individual and verify that the table updates accordingly

## Note:
Test it across different user types, and let me know if I missed any expected conventions or if you have any other suggestions.

### CODE REVIEW

**Review by manikittu810 — COMMENTED** [review]
Hello @yashwanth170 
 I have tested this , and it is displaying the text Loading for a long time instead of a table and I hjave also tried using clearing the cache.
Screenshot 2024-11-14 at 5 52 42 PM

Thank you.

**Comment by yashwanth170** [comment]
> Hello @yashwanth170 I have tested this , and it is displaying the text Loading for a long time instead of a table and I hjave also tried using clearing the cache. Screenshot 2024-11-14 at 5 52 42 PM
> 
> Thank you.

Hello,
Did you checkout the backend branch as well?

File: src/components/PermissionsManagement/UserPermissionsPopUp.jsx [hunk 1]
```diff
@@ -94,6 +135,7 @@ function UserPermissionsPopUp({
             autoClose: 10000,
           });
           setToastShown(true);
+          logPermissionChanges(logPermissionChangesUrl, permissionChangeLog);
```
Comment (nathanah) [inline src/components/PermissionsManagement/UserPermissionsPopUp.jsx:138]:
```suggestion
```
This is not atomic. This request could fail after the permission change request succeeds, so it wouldn't show up in the history. If a dev wanted to test updating permissions manually with something like Postman, they could avoid also logging  those changes which could cause confusion or potential problems.
Logging permission changes should be part of the same API call as updating the user's permissions.

File: src/utils/URL.js [hunk 2]
```diff
@@ -5,6 +5,8 @@ const APIEndpoint =
 export const ENDPOINTS = {
   APIEndpoint: () => APIEndpoint,
   USER_PROFILE: userId => `${APIEndpoint}/userprofile/${userId}`,
+  POST_USER_PERMISSION_CHANGE_LOGS: `${APIEndpoint}/logPermissionChanges`,
```
Comment (nathanah) [inline src/utils/URL.js:8]:
```suggestion
```

File: src/components/PermissionsManagement/UserPermissionsPopUp.jsx [hunk 3]
```diff
@@ -80,9 +90,40 @@ function UserPermissionsPopUp({
...
+      authUser: {
+        email: authUser.email,
+        role: authUser.role,
+      },
+      userId,
+      existingPermissions, // Existing permissions before change
+      addedPermissions, // New permissions added
+      removedPermissions, // Permissions removed
+    };
+
```
Comment (nathanah) [inline src/components/PermissionsManagement/UserPermissionsPopUp.jsx:126]:
```suggestion
```
This logic should be on the backend and part of the same API call as the update itself.

File: src/components/PermissionsManagement/UserPermissionsPopUp.jsx [hunk 4]
```diff
@@ -80,9 +90,40 @@ function UserPermissionsPopUp({
     const userId = actualUserProfile?._id;
 
     const url = ENDPOINTS.USER_PROFILE(userId);
+    const logPermissionChangesUrl = ENDPOINTS.POST_USER_PERMISSION_CHANGE_LOGS;
```
Comment (nathanah) [inline src/components/PermissionsManagement/UserPermissionsPopUp.jsx:93]:
```suggestion
```

File: src/components/PermissionsManagement/UserPermissionsPopUp.jsx [hunk 5]
```diff
@@ -64,6 +64,16 @@ function UserPermissionsPopUp({
...
+  const logPermissionChanges = async (logPermissionChangesUrl, permissionChangeLog) => {
+    try {
+      await axios.post(logPermissionChangesUrl, permissionChangeLog);
+    } catch (error) {
+      toast.error('Error logging permissions:', {
+        autoClose: 10000,
+      });
+    }
+  };
+
```
Comment (nathanah) [inline src/components/PermissionsManagement/UserPermissionsPopUp.jsx:76]:
```suggestion
```

File: src/components/PermissionsManagement/PermissionsManagement.jsx [hunk 6]
```diff
@@ -243,7 +244,7 @@ function PermissionsManagement({ roles, auth, getUserRole, userProfile, darkMode
       </div>
       {loading && <p className="loading-message">Loading...</p>}
       {changeLogs?.length > 0 && (
-        <PermissionChangeLogTable changeLogs={changeLogs.slice().reverse()} darkMode={darkMode} />
+        <PermissionChangeLogTable changeLogs={changeLogs} darkMode={darkMode} />
```
Comment (nathanah) [inline src/components/PermissionsManagement/PermissionsManagement.jsx:247]: removing reverse?

**Review by nathanah — CHANGES_REQUESTED** [review]
Most of this logic should be on the backend. The only frontend changes required should be accommodating the new info in the `permissionChangeLogTable`.

**Review by audreydieuanh — CHANGES_REQUESTED** [review]
The table doesn't show, and I have done git checkout for backend PR #1149.
test11

**Review by nathanah — CHANGES_REQUESTED** [review]
Not sure if this should be on the front-end or during logging on the back-end, but individuals should be labeled as "INDIVIDUAL: <Name>" according to instruction C. in your screenshot. Everything else looks good.
image

**Comment by nathanah** [comment]
> The table doesn't show, and I have done git checkout for backend PR #1149. ![test11](https://private-user-

You have to be logged into an Owner account to see the table.

**Review by nathanah — APPROVED** [review]

**Comment by one-community** [comment]
Thank you all, merging!

**Pull request merged**